firebase/rules/tests/invocations.spec.ts

```diff
@@ -0,0 +1,161 @@
+import 'jest';
+import * as matchers from '../matchers';
+import * as targaryen from 'targaryen';
+
+import { SYSTEM_USER, Invocation, InvocationType } from '@machinelabs/models';
+
+const anonymousUser = targaryen.util.users.anonymous;
+const rules = require('../database.rules.json');
+
+const dummyInvocation: Invocation = {
+  id: '1',
+  timestamp: Date.now(),
+  user_id: null,
+  type: Date.now() % 2 ? InvocationType.StartExecution : InvocationType.StopExecution,
+  data: null
+};
+
+const dummyInvocationLab = {
+  id: '1',
+  directory: 'dummy/directory'
+};
+
+const dummyInvocationExecution = {
+  execution_id: '1',
+};
+
+describe('/invocations', () => {
+  let currentUser;
+  let systemUser;
+
+  let dummyInvocationCurrentUser: Invocation;
+  let dummyInvocationSystemUser: Invocation;
+
+  beforeEach(() => {
+    jest.addMatchers(matchers);
+
+    targaryen.util.setFirebaseData({
+      users: {
+        '1': {
+          common: {
+            id: '1',
+            displayName: 'Foo Bar',
+            email: 'foo.bar@example.com',
+            bio: '',
+            isAnonymous: false,
+            photoUrl: 'avatars.com/1'
+          }
+        },
+        '2': {
+          common: {
+            id: '2',
+            displayName: 'Fizz Buzz',
+            email: 'fizz.buzz@example.com',
+            bio: '',
+            isAnonymous: false,
+            photoUrl: 'avatars.com/2'
+          }
+        }
+      },
+      invocations: {
+        '1': {
+          common: {
+            id: '1',
+            timestamp: Date.now(),
+            user_id: '1',
+            type: InvocationType.StartExecution,
+            data: dummyInvocationExecution
+          }
+        },
+        '3': {
+          common: {
+            id: '3',
+            timestamp: Date.now(),
+            user_id: '2',
+            type: InvocationType.StopExecution,
+            data: dummyInvocationLab
+          }
+        }
+      },
+      idx: {
+        invocation_rate_proof: {
+          '1': {
```

Comment:
@maartentibau  By fixing the rake proofs and the invocation id in the test data you gave the tests a free ticket to satisfy the rate proofs. It would be great to actually have a test that proofs that a user is NOT allowed to write an invocation if it doesn't satisfy the rate proof.

The API rate limits (1 API call per second per user) is an important aspect of keeping the system secure against attacks so it would be great to have tests for that as well.

Reply:
OK, trying to understand what you mean, because I wasn't aware about the API rate limits to be honest.

So you want me to write a test that fires a `write` to an envocation every 500ms and that should fail?

Reply:
You can not have this fully dynamic since the tests run against static data. So you don't need to actually emulate multiple requests.

However you should be able to prove that you can not write an invocation with an `id` that doesn't match the key of the rate proof.

This is exactly what the `write` rule also [checks](https://github.com/machinelabs/machinelabs/blob/33d37a630fe7a66efd2058f6734f4ca850820ba1/firebase/rules/database.rules.bolt#L207).

```
write() { (isOwningUser(this, 'user_id') && isInvocationRateProofKey(auth.uid, $id)) || isSystem() }
```

Reply:
I think this test proves what you're saying here, no?

```
it('Only system users can write into invocations of other users', () => {
  expect(anonymousUser).cannotWrite('/invocations/3/common', dummyInvocation);
  expect(currentUser).cannotWrite('/invocations/3/common', dummyInvocationCurrentUser);
  expect(systemUser).canWrite('/invocations/3/common', dummyInvocationSystemUser);
});
```

Reply:
You can test this by having a user that *does* have write access, but no invocationRateProof. For example:

```
invocation: {
  id: 53
  ...
}
```

Can only be written if:

```
invocation_rate_proof: {
  24: {
    key: 53
   }
}
```

^ exists and the auth user id === 24 **and** the `key` === the invocation

So this means you could setup a dummy invocation with an id that doesn't exist in any invocation rate proof entry

Reply:
@PascalPrecht that's what I do in my last test. I do a write, on the invocation with id = 3.
My userId === 1, that user only has write access to the invocation with id === 1 and not 3.

Reply:
OK, will do that

Reply:
@PascalPrecht big up for the crystal clear instructions. :heart:

@maartentibau Sorry my initial explanation was rather terse. Thanks you for biting your way through it!

Reply:
@PascalPrecht 

Added an extra test, but I changed the text a bit. Because it's actually only a system user that can write when the rate proof is unsatisfied. Hope this is what you were searching for.